server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java

```diff
@@ -214,44 +214,117 @@ public BytesReference build() throws IOException {
+                sb.append(metric);
+            }
+            final BytesReference metricFieldsHash = hash128(new BytesRef(sb.toString()));
+
+            // NOTE: catenate all dimension value hashes up to names certain number of dimensions
+            int numberOfDimensions = Math.min(MAX_DIMENSIONS, dimensions.size());
+            int buf1Index = 0;
+            final byte[] buf1 = new byte[4 * numberOfDimensions];
+            for (final DimensionDataHolder dimension : dimensions) {
+                if (buf1Index >= buf1.length) break;
+                final BytesRef dimensionValueBytesRef = dimension.value().toBytesRef();
+                ByteUtils.writeIntLE(StringHelper.murmurhash3_x86_32(dimensionValueBytesRef, 0), buf1, buf1Index);
+                buf1Index += 4;
+            }
+            final BytesRef valueByValue = new BytesRef(buf1, 0, buf1.length);
+
+            // NOTE: hash all dimension field allValues
+            int buf2Index = 0;
+            final byte[] buf2 = new byte[dimensionValuesLength];
+            for (final DimensionDataHolder dimension : dimensions) {
+                final BytesRef dimensionValueBytesRef = dimension.value.toBytesRef();
+                System.arraycopy(
+                    dimensionValueBytesRef.bytes,
+                    dimensionValueBytesRef.offset,
+                    buf2,
+                    buf2Index,
+                    dimensionValueBytesRef.length - dimensionValueBytesRef.offset
+                );
+                buf2Index += (dimensionValueBytesRef.length - dimensionValueBytesRef.offset);
+            }
+            final BytesReference dimensionValuesHash = hash128(new BytesRef(buf2, 0, buf2.length));
+
+            final BytesRef names = dimensionNamesHash.toBytesRef();
+            final BytesRef metrics = metricFieldsHash.toBytesRef();
+            final BytesRef allValues = dimensionValuesHash.toBytesRef();
+            int tsidLength = names.length + metrics.length + valueByValue.length + allValues.length;
+            byte[] tsid = new byte[tsidLength];
+            int tsidIndex = 0;
+            System.arraycopy(names.bytes, names.offset, tsid, tsidIndex, names.length);
+            tsidIndex += names.length;
+            System.arraycopy(metrics.bytes, metrics.offset, tsid, tsidIndex, metrics.length);
+            tsidIndex += metrics.length;
+            System.arraycopy(valueByValue.bytes, valueByValue.offset, tsid, tsidIndex, valueByValue.length);
+            tsidIndex += valueByValue.length;
+            System.arraycopy(allValues.bytes, allValues.offset, tsid, tsidIndex, allValues.length);
+            tsidIndex += allValues.length;
+            assert tsidIndex == tsidLength;
+            try (BytesStreamOutput out = new BytesStreamOutput()) {
+                out.writeVInt(TSID_HASH_SENTINEL);
+                out.writeBytesRef(new BytesRef(tsid, 0, tsidLength));
+                return out.bytes();
+            }
```

Comment:
Do we need the hash sentinel?
Can we simplify this to `return new BytesArray(tsid)`?

Reply:
The hash sentinel is not really needed...I will remove it once I address removing the `decodeTsid` method which is the only place where it is used.  After hashing, indeed, we lose the ability to decode the tsid field and have to go through doc values.

Reply:
I still didn;t do that because not being able to use `decodeTsid` (unfortunately) breaks downsampling.